client/src/_transport_layer.rs

```diff
@@ -0,0 +1,257 @@
+    forward_transfer: HashMap<NodeId, ForwardSender>,
+    forward_reliable: HashMap<NodeId, ForwardSender>,
+}
+
+impl TransportLayer {
+    pub fn new(config: Arc<ClientConfig>) -> TransportLayer {
+        let out = Channel::<Forwarded>::default();
+        let session_layer = SessionLayer::new(config.clone());
+        let virtual_tcp = TcpLayer::new(
+            &config.node_pub_key,
+            &config.stack_config,
+            &out,
+            session_layer.clone(),
+        );
+
+        TransportLayer {
+            config,
+            session_layer,
+            virtual_tcp,
+            state: Arc::new(RwLock::new(TransportLayerState {
+                forward_unreliable: Default::default(),
+                forward_transfer: Default::default(),
+                forward_reliable: Default::default(),
+            })),
+            ingress_channel: out,
+        }
+    }
+
+    pub(crate) async fn spawn(&mut self) -> anyhow::Result<SocketAddr> {
+        let bind_addr = self.session_layer.spawn().await?;
+        self.virtual_tcp
+            .spawn(self.session_layer.config.node_id)
+            .await?;
+
+        self.spawn_ingress_handler().await?;
+        Ok(bind_addr)
+    }
+
+    pub async fn shutdown(&mut self) -> anyhow::Result<()> {
+        let channels = {
+            let mut state = self.state.write().await;
+
+            let channels1 = state.forward_unreliable.drain().collect::<Vec<_>>();
+            let channels2 = state.forward_transfer.drain().collect::<Vec<_>>();
+            let channels3 = state.forward_reliable.drain().collect::<Vec<_>>();
+
+            channels1
+                .into_iter()
+                .chain(channels2.into_iter())
+                .chain(channels3.into_iter())
+                .collect::<Vec<_>>()
+        };
+        for (_, mut channel) in channels {
+            channel.disconnect().await.ok();
+        }
+
+        self.virtual_tcp.shutdown().await;
+
+        // After Tcp shutdown will return, we are sending last Tcp packet to notify other Node,
+        // that connection is closed. We shouldn't close sessions before we give them chance to be sent.
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        self.session_layer.shutdown().await
+    }
+
+    pub fn forward_receiver(&self) -> Option<ForwardReceiver> {
+        self.ingress_channel.receiver()
+    }
+
+    async fn dispatch(&self, packet: Forwarded) {
+        log::trace!(
+            "[TransportLayer] Dispatching packet from [{}] ({})",
+            packet.node_id,
+            packet.transport
+        );
+
+        match &packet.transport {
+            TransportType::Unreliable => self.dispatch_unreliable(packet).await,
+            TransportType::Reliable => self.virtual_tcp.dispatch(packet).await,
+            // We shouldn't get this transport type from `SessionLayer`, because only TcpLayer
+            // can distinguish packets between `Reliable` and `Transfer`.
```

Comment:
It's the opposite: `SessionLayer` never returns `TransportType::Transfer`

Reply:
```suggestion
            // Currently `SessionLayer` responds only with `Unreliable` and `Reliable`, because only TcpLayer
            // can distinguish packets between `Reliable` and `Transfer`.
            // Nevertheless this function will work correctly even when getting `Transfer` variant.
```